oauth2client/django_util/decorators.py

```diff
@@ -0,0 +1,50 @@
+from django.shortcuts import redirect
+from oauth2client.django_util.storage import get_storage
+from six import wraps
+from . import settings
+
+
+def _build_scope_query_param(scopes):
+    s = ''
+    for scope in scopes:
+        s += '&scopes=%s' % scope
+    return s
+
+
+def required(decorated_function=None, scopes=None, **decorator_kwargs):
+    def curry_wrapper(wrapped_function):
+        @wraps(wrapped_function)
+        def required_wrapper(*args, **kwargs):
+            request = args[0]
+            assert hasattr(request, 'session'), (
+                "The Google OAuth2 Helper requires session middleware "
+                "to be installed. Edit your MIDDLEWARE_CLASSES setting to include "
+                "'django.contrib.sessions.middleware.SessionMiddleware'."
+            )
+
+            return_url = decorator_kwargs.pop('return_url', request.build_absolute_uri())
+            storage = get_storage(request)
+            credentials = storage.get()
+
+            requested_scopes = list(settings.GOOGLE_OAUTH2_SCOPES) + (scopes or [])
+            if credentials:
+                requested_scopes += list(credentials.scopes)
+
+            # If no credentials or if existing credentials but mismatching
+            # scopes, redirect for incremental authorization.
+            if not credentials or not credentials.has_scopes(requested_scopes):
+                response = redirect('google_oauth:authorize')
```

Comment:
See my comment above about `reverse`, modifying the location after doing this seems really hacky.